### ISSUE TITLE
bootctl accepts relative paths for linux and initrd

### ISSUE THREAD

**Issue by ebsf1** [systemd/systemd#23913]
### systemd version the issue has been seen with

249

### Used distribution

Ubuntu 22.04

### Linux kernel version used

5.15.0-40-generic; 5.15.0-37-generic

### CPU architectures issue was seen on

x86_64

### Component

bootctl

### Expected behaviour you didn't see

Boot entries conforming with the [Boot Loader Specification](https://systemd.io/BOOT_LOADER_SPECIFICATION/) according to (a) its language; and (b) the output of `bootctl list`, which according to its [documentation](https://www.freedesktop.org/software/systemd/man/bootctl.html#list) returns conforming entries, would be observed and appear in the list on the systemd-boot screen.

### Unexpected behaviour you saw

Conforming entries do not appear in the systemd-boot list.

Specifically, systemd-boot, perhaps relying on systemd-bootx64.efi, does not recognize paths to kernel files on the file system of the EFI mount point or otherwise not on the ESP, _e.g._, ../vmlinuz or ../initrd.img.

It is elementary that the entire purpose of any boot loader (and therefore systemd-boot, perhaps in conjunction with systemd-bootx64.efi), after all, is simply to hand off execution to kernel files on another partition and file system.  systemd-boot seems incapable, however, of reading them.

Indeed, the entire construct of `kernel-install` and its related `bootctl` options, appears designed to compensate for the absence of this capability.  It attempts too much.  Das Gummiboot wird gesunken.  Establish this fundamental read capability and none of this is necessary.  Kernel files can remain encapsulated on the partitions they boot and kernel installation and update procedures can be left intact.

### Steps to reproduce the problem

This is a multiboot Linux system, chiefly with different versions and flavours of Ubuntu installed on discrete partitions, all on a single volume.  The ESP is /dev/sda1, /dev/sda2-7 are bootable, /dev/sda8 is swap, and /dev/sda9 is data.  The installations were competing for control over the GRUB boot loader, prompting an attempt to replace it with systemd-boot.

The base configuration is a conventional Ubuntu Desktop 22.04 installation:
- Kernel files are located in /boot, along with the symlinks vmlinuz and initrd.img, which reference the latest versions.
- The ESP is mounted to /boot/efi via fstab.

/boot/efi/loader/loader.conf is:

```
#	/boot/efi/loader/loader.conf
#	systemd-boot configuration file

timeout 5
#default [machine-id redacted]-*
#	From man loader.conf, Options, default, at https://www.freedesktop.org/software/systemd/man/loader.conf.html:
#default @saved
default sdx4.conf
console-mode keep
```

/boot/efi/loader/entries/sdx3.conf is:

```
#	/boot/efi/loader/entries/sdx3.conf
#	Boot entry configuration file.

title sdx3 - Ubuntu Desktop 22.04
linux ../vmlinuz
initrd ../initrd.img
options root=PARTUUID=[PARTUUID redacted] rw
```

`# bootctl list` returns:

```
        title: sdx3 - Ubuntu Desktop 22.04
           id: sdx3.conf
       source: /boot/efi/loader/entries/sdx3.conf
        linux: ../vmlinuz
       initrd: ../initrd.img
      options: root=PARTUUID=[PARTUUID redacted] rw

        title: sdx4 - Ubuntu Desktop 22.04
           id: sdx4.conf
       source: /boot/efi/loader/entries/sdx4.conf
        linux: /vmlinuz (No such file or directory)
       initrd: /initrd.img (No such file or directory)
      options: root=PARTUUID=[PARTUUID redacted] rw
```

So it is that we see that specifying the kernel files' location as being in the parent directory of the EFI System Partition's mount point by using the ../ path conforms not only with the language of the Boot Loader Specification (which I should hasten to add contains no language requiring kernel files to be located on the ESP) but also with the code that implements it.

**Comment by medhefgo** [comment]
> So it is that we see that specifying the kernel files' location as being in the parent directory of the EFI System Partition's mount point by using the ../ path conforms not only with the language of the Boot Loader Specification (which I should hasten to add contains no language requiring kernel files to be located on the ESP) but also with the code that implements it.

You really ought to read the spec. It literally says "`linux` is the Linux kernel to spawn and as a path ***relative to file system root***". And further up it says what is meant with file system root (hint: the *ESP* or XBOOT). And I sincerely hope you do not expect sd-boot to pick up these files from the filesystem on `/boot`?!?

And you also should be reading the output of `bootctl`: It *literally* says "No such file or directory" for the sdx4 (again, you cannot expect a bootloader to understand your arbitrary mount hierarchy from linux userspace). If bootctl cannot find your kernel, how is sd-boot supposed to?!?

At best this is a bug in `bootctl` no enforcing this when looking whether files exist.

**Comment by medhefgo** [comment]
> At best this is a bug in `bootctl` no enforcing this when looking whether files exist.

A simple test file confirms this:
```
title Test
linux ../bin/sh
```

**Comment by poettering** [comment]
> Specifically, systemd-boot, perhaps relying on systemd-bootx64.efi, does not recognize paths to kernel files on the file system of the EFI mount point or otherwise not on the ESP, _e.g._, ../vmlinuz or ../initrd.img.

The spec is very explicit:

> When Type #1 configuration snippets refer to other files (for linux, initrd, efi, devicetree, and devicetree-overlay), those files must be located on the same partition, and the paths must be absolute paths relative to the root of that file system. The naming of those files can be chosen by the installer. A recommended scheme is described in the next section.

https://systemd.io/BOOT_LOADER_SPECIFICATION/

**Comment by poettering** [comment]
Note that from the boot loader perspective in the UEFI world there's no /boot or /boot/efi mount point or such. Instead there are just partitions: the ESP one, and possibly others, and they have not UNIX mount point, because there is no UNIX. Hence there is also no "../" relationship.

Note that sd-boot actually searches for boot loader entries in two partitions: the ESP and XBOOTLDR. But they must both be in a format that UEFI can acccess, i.e. VFAT. You could thus make your /boot/ partition XBOOTLDR. That said, I#d not bother with the complexity and just put everything in the ESP. Simpler, more robust.

**Comment by ebsf1** [comment]
@medhefgo and @poettering, I regret that you did not apprehend that I parsed the spec especially closely.  I would not have posted otherwise.

And yes, I do actually expect sd-boot (or its boot loader) to execute kernels located on the partitions they boot.  That seems to be the very purpose of a boot loader and, for example, the GRUB boot loaders appear entirely capable of it.  Consequently, such a notion does not seem to break new ground.  Furthermore, the syntax checker, `bootctl list`, confirms the hypothesis notwithstanding that this now is characterized as a bug elsewhere.

This said, I am only an autodidact in computer science.  I am trained as a lawyer and editor (and financier).  I appreciate that the intent of the spec may be as you describe.  Nevertheless, it seems to have been written by someone with a more conversational than rigorous grasp of English grammar, is ambiguous or vague in many respects, and simply does not literally convey the meaning you assert.  Articles are missing, pronoun antecedents are absent, verb tense often is incorrect, and careless use of the passive voice often dilutes clarity.  Consequently, much of the usage is at best indirect.

I take the points you make, and certainly insofar as intent is concerned.  Let me address those points, not to be argumentative or personally critical but for the sake of a rigorous understanding of the language and the meaning it does, and does not, convey.

The statement about being a path relative to file system root only addresses the syntax of the value for the `linux` key.  It says nothing about where the specified kernel must be.  The statement also is ambiguous because the prepositional clause, besides lacking an article, does not specify which file system, _i.e._, its antecedent is absent.  The statement also does not negate paths to filesystems on which the ESP root is mounted.  Note also that even if it did, such a statement still would not be an express definition of where kernel files can, must, or cannot be, not only grammatically but also because of the potential of bind mounts.

Concerning the sdx4 entry, there simply were no kernel files installed at the time in the paths specified, _i.e._, ESP root, which accounts for the "no such file or directory" error.  I believe, for that partition, they were in a `machine-id/version` subdirectory at that point.  Regardless, the point of including that entry was to contrast the other entry, in which the kernel files were located in /boot and `bootctl list` validated the `../` syntax and therefore confirmed that the entry was consistent with the spec.

The statement about Type #1 configuration snippets is fraught in several respects.  The first clause only states that other files must be located on the same partition.  The statement does not say which partition; again, the antecedent is absent.  Grammatically, this statement only requires other files not to be distributed among many partitions.  It does not negate locating other files on non-ESP partitions.  What's missing is an "as" clause; all that is implied is "as one another."

The second clause, concerning path specification, is consistent with the language for the `linux` key syntax and so both expresses the same meaning and suffers the same difficulties.  It also is ambiguous in two ways:  The usage "that filesystem" doesn't express which filesystem.  That is, the antecedent of the indefinite relative pronoun "that" is absent.  It could be the filesystem of the ESP or of the partition on which the other files are located.  This ambiguity stands on top of the ambiguous partition reference.  We can surmise, albeit without certainty, that partitions and filesystems probably are meant to coincide but the meaning of this language (notwithstanding the use of imperatives ("must")) nevertheless becomes more convoluted the more closely it is examined.  The usage here fails in many respects.

Beyond this, the spec contains no active, affirmative statement of where the kernel files can, must, or cannot be.

I could go on but the essential point is that the spec's language isn't up to the same standards as the software and consequently does not literally convey the meaning you assert.  In fact, it supports my hypothesis, and more so the closer one reads.

Again, I'm not debating the intended meaning, sd-boot, its code, or architecture.  I also understand the natural inclination to debate what the language "must" mean, but this really reduces to either an assertion of subjectively intended meaning, or an admission that the asserted meaning is merely implied.  Clear writing requires objectivity and to anticipate one's audience.  The spec, however, seems to have been written by the developers, however unconsciously, for themselves.  In a sense, this would be doubly subjective, the audience coinciding with the authors, who developed the code.  Regardless, sd-boot simply does not implement the spec as written.

If it's seen to be constructive, I could make some revisions for your consideration to improve the grammar, usage, and clarity of the document.  Please let me know whether and how you'd like to proceed.

**Comment by poettering** [comment]
> The statement about being a path relative to file system root only addresses the syntax of the value for the `linux` key. It says nothing about where the specified kernel must be.

It does not? "…must be located on the same partition…" is what the spec says. See quote above.

**Comment by poettering** [comment]
#24541 will now validate paths from bootspec snippets more strictly and warn about those which contain ".."

#24521 contains various updates to the boot loader spec. It explicitly makes clear ".." is not OK. It already said that files referenced from snippets must be on the same partition, see above, so it doesn't change anything there.

Updates to the spec are certainly welcome, but do note the spec has a technical focus (not a legal one), and is supposed to be digestable by regular hackers, and hence relatively terse and not too redundant.

If you decide to work on the spec, please file a new PR on it. But please base your work on #24521 once it is merged to minimize merge conflicts.

**Comment by ebsf1** [comment]
> #24541 will now validate paths from bootspec snippets more strictly and warn about those which contain ".."
> 
> #24521 contains various updates to the boot loader spec. It explicitly makes clear ".." is not OK. It already said that files referenced from snippets must be on the same partition, see above, so it doesn't change anything there.
> 
> Updates to the spec are certainly welcome, but do note the spec has a technical focus (not a legal one), and is supposed to be digestable by regular hackers, and hence relatively terse and not too redundant.
> 
> If you decide to work on the spec, please file a new PR on it. But please base your work on #24521 once it is merged to minimize merge conflicts.

It isn't a question of legal or technical focus, it's a question of clear writing.  My legal background simply allows me to approach writing with rigor.  I'm a regular hacker, without any formal CS training.  I'm the audience, and the spec does / did not answer many basic points.  It is ambiguous, roundabout at best, and hardly terse.  Ultimately, it fails / failed as a technical document because it fails / failed as a document.

I'll wait for the next rev of the spec to publish and then have a look.  I'll take a crack at the Github mechanism but haven't used it, so you may be left with a redlined WordPerfect document.  Stay tuned.  I have business to run and this trainwreck put me back a full month.  It will be a fairly comprehensive overhaul if I can find the time, just because it needs it.

In the meantime, get to work on relative paths (i.e., system root paths or paths relative to ESP root).  Above all else, a boot loader must be able to boot a partition.  sd-boot fails, and profoundly, because it cannot, and that is because it cannot read kernel files on the partition to be booted.  The mind boggles.  It is astonishing that this has not existed from the outset.  Quibbles about where is best for kernel files are just that-quibbles.  What's important are the basics.

The sd-boot value proposition is a single immutable EFI boot loader that accepts configuration from simple entry configuration files.  That's all that anyone cares about.  That needs to rest on the ability to read the kernel files where they already are, however.  (That location already is a parameter, and it's okay if the supported filesystems are limited based on adoption (i.e., ext4, NTFS, and zfs).)  Without that, it's just a stupid, irrelevant software project.

**Comment by vans163** [comment]
Is `not-relative-to-/boot` paths not supported? This breaks a enclave usecase where the initramfs and kernel are encrypted until TDX/SEV decrypt via EFI module.

**Comment by medhefgo** [comment]
> Is `relative-to-/boot` paths not supported? This breaks a secure compute usecase where the initramfs and kernel are sitting on encrypted disk in /boot waiting until TDX/SEV decrypt the disk via EFI module.

You'll have to be more specific, because I have no idea how that stuff works (especially how the trust is bootstrapped).

Files referenced by boot config entries must be relative to the filesystem root (ESP or XBOOTLDR) the config resides in (and therefore by extension must be accessible/readable by the firmware/bootloader).

**Comment by vans163** [comment]
> > Is `relative-to-/boot` paths not supported? This breaks a secure compute usecase where the initramfs and kernel are sitting on encrypted disk in /boot waiting until TDX/SEV decrypt the disk via EFI module.
> 
> You'll have to be more specific, because I have no idea how that stuff works (especially how the trust is bootstrapped).
> 
> Files referenced by boot config entries must be relative to the filesystem root (ESP or XBOOTLDR) the config resides in (and therefore by extension must be accessible/readable by the firmware/bootloader).

I am not sure why this is required as GRUB does not require it. Is there a specific reason for this, because it exposes the kernel and initramfs in cleartext in this scenario.

**Comment by medhefgo** [comment]
I fail to see what this is supposed to protect against. There is no reason for kernel+initrd to contain any sensitive information. In fact, UKIs are moving to a built-by-the-vendor model, so anyone can just look at your bootloader and figure out what kernel+initrd you'll be using.

To access other filesystems (especially encrypted ones) you need to provide a ton of code for that with rather little gain. All probably written in c to access a device that contains sensitive data. EFI isn't a particularly nice environment to write that kind of stuff. We already struggle enough with all kinds of firmware quirks, having to implement custom fs/encryption drivers on top of that is just not a good idea. There is also the matter that the code won't be properly optimized (even sse has to be enabled before it can be used), and we don't want to deal with that complication either. Not to mention, you'll want to deal with keyboard layouts, different languages for the UI, automatic unlock with TPM, etc…

You know a project that has all that already? It's called the linux kernel (along with an initrd). Just use a UKI with TPM to get your trust properly established. Or, you know, nobody's stopping you from using grub.

**Comment by vans163** [comment]
Its better to keep it encrypted if possible. There is no reason for it to contain sensitive information but it still gives away what is booting.

It does not matter what UKIs are doing. If UKI jumped off a bridge, would .. ?

Not really, your not accessing it encrypted, your doing your attestation with the secure element before the initramfs is even loaded/accessed via EFI module.

It does not matter if EFI isn't a nice environment.

Struggle is real but no custom encryption/fs is implemented, simply the initramfs is loaded as normal as the boot process continues, which it continues after the trust has been established and the disk has been decrypted.

```
part1 (cleartext)
/boot/efi/

part2 (encrypted)
/
/boot/vmlinux
/boot/initrd
```
Basically vmlinux and initrd lye on part2 encrypted, waiting for attestation+decryption. Then boot continues.

You can see a reference implementation here https://blog.hansenpartnership.com/category/security/confidential-computing/, but its dated already TDX is getting upstreamed in 6.2 likely.

TPM is irrelevent.

Indeed no one is stopping from using GRUB but secure compute use case will become mainstream in a few years.
